src/holochain/admin_conn.rs

```diff
@@ -1,47 +1,154 @@
+    /// to the specified socket address.
+    pub fn new(socket_addr: SocketAddr) -> Self {
+        Self {
+            socket_addr,
+            connection_handle: Default::default(),
+        }
+    }
+
+    /// Allows calling a method on the [AdminWebsocket], with automatic reconnection if needed
+    pub async fn call<T>(
+        &self,
+        execute: impl Fn(AdminWebsocket) -> BoxFuture<'static, HcHttpGatewayResult<T>>,
+    ) -> HcHttpGatewayResult<T> {
+        for _ in 0..=ADMIN_WS_CONNECTION_MAX_RETRIES {
+            let admin_ws = self.get_admin_ws().await?;
+
+            match execute(admin_ws).await {
+                Ok(output) => return Ok(output),
+                Err(HcHttpGatewayError::HolochainError(ConductorApiError::WebsocketError(e))) => {
+                    tracing::warn!(
+                        ?e,
+                        "Detected admin websocket disconnection. Attempting to reconnect"
+                    );
+                    *self.connection_handle.write().await = None;
+                    continue;
+                }
+                Err(e) => return Err(e),
+            }
+        }
 
-/// Placeholder for the admin connection.
-#[derive(Debug)]
-pub struct AdminConn;
+        Err(HcHttpGatewayError::UpstreamUnavailable)
+    }
+
+    async fn get_admin_ws(&self) -> HcHttpGatewayResult<AdminWebsocket> {
+        {
+            let lock = self.connection_handle.read().await;
+
+            if let Some(admin_ws) = lock.as_ref() {
+                return Ok(admin_ws.clone());
+            }
+        }
+
+        let mut lock = self.connection_handle.write().await;
+
+        match AdminWebsocket::connect(self.socket_addr).await {
+            Ok(admin_ws) => {
+                tracing::info!("Connected a new Holochain admin websocket");
+                *lock = Some(admin_ws.clone());
+                Ok(admin_ws)
+            }
+            Err(e) => {
+                tracing::error!(?e, "Failed to connect Holochain admin websocket");
+                Err(HcHttpGatewayError::UpstreamUnavailable)
+            }
+        }
+    }
+}
 
 impl AdminCall for AdminConn {
     fn list_app_interfaces(
         &self,
     ) -> BoxFuture<'static, HcHttpGatewayResult<Vec<AppInterfaceInfo>>> {
-        todo!()
+        let this = self.clone();
+        Box::pin(async move {
+            this.call(|admin_ws| Box::pin(async move { Ok(admin_ws.list_app_interfaces().await?) }))
+                .await
+        })
     }
 
     fn issue_app_auth_token(
         &self,
-        _payload: IssueAppAuthenticationTokenPayload,
+        payload: IssueAppAuthenticationTokenPayload,
     ) -> BoxFuture<'static, HcHttpGatewayResult<AppAuthenticationTokenIssued>> {
-        todo!()
+        let this = self.clone();
+        Box::pin(async move {
+            this.call(|admin_ws| {
+                let payload = IssueAppAuthenticationTokenPayload {
```

Comment:
We should probably make this clone, but that doesn't justify doing a release of 0.4 so for now this will do. A change we should make on 0.5 I think